slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,316 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// PodE2EStartupLatencyKey is a key for pod startup latency monitoring metric.
+	PodE2EStartupLatencyKey = "slomonitor_pod_e2e_startup_latency_seconds"
+	// Because for some reason it's not present in versioned api...
+	NamespaceSystem = "kube-system"
+)
+
+var (
+	// PodE2EStartupLatency is a prometheus metric for monitoring pod startup latency.
+	PodE2EStartupLatency = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    PodE2EStartupLatencyKey,
+			Help:    "Pod e2e startup latencies in milliseconds, without image pull times",
+			Buckets: prometheus.LinearBuckets(0.5, 0.25, 50),
+		},
+	)
+)
+
+var registerMetrics sync.Once
+
+// Register registers all prometheus metrics.
+func Register() {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(PodE2EStartupLatency)
+	})
+}
+
+// PodStartupMilestones keeps all milestone timestamps from Pod creation.
+type PodStartupMilestones struct {
+	created          time.Time
+	startedPulling   time.Time
+	finishedPulling  time.Time
+	observedRunning  time.Time
+	deletedTimestamp time.Time
+	seenPulled       int
+	needPulled       int
+	accountedFor     bool
+}
+
+// PodStartupLatencyDataMonitor monitors pod startup latency and exposes prometheus metric.
+type PodStartupLatencyDataMonitor struct {
+	sync.Mutex
+	kubeClient     clientset.Interface
+	PodStartupData map[string]PodStartupMilestones
+	// Slice of strings marked for deletion waiting to be purged. Just an optimization to avoid
+	// full scans.
+	toDelete []string
+	// Time after which deleted entries are purged.
+	purgeAfter time.Duration
+}
+
+func (data *PodStartupMilestones) IsComplete() bool {
+	return !data.created.IsZero() && !data.startedPulling.IsZero() && !data.finishedPulling.IsZero() && !data.observedRunning.IsZero() && data.seenPulled == data.needPulled && !data.accountedFor
+}
+
+// NewPodStartupLatencyDataMonitor creates a new PodStartupLatencyDataMonitor from a given client.
+func NewPodStartupLatencyDataMonitor(c clientset.Interface, purgeAfter time.Duration) *PodStartupLatencyDataMonitor {
+	return &PodStartupLatencyDataMonitor{
+		kubeClient:     c,
+		PodStartupData: map[string]PodStartupMilestones{},
+		purgeAfter:     purgeAfter,
+	}
+}
+
+// Run starts a PodStartupLatencyDataMonitor: it creates all watches, populates PodStartupData and updates
+//  PodE2EStartupLatency metric
+func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
+	controller := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				obj, err := pm.kubeClient.Core().Pods(NamespaceSystem).List(options)
```

Comment:
Now when I think about it, I think that doing the hack with NamespaceSystem might be pretty misleading. We should either clearly document it, or get back to NamespaceAll here (sorry for going back and forth on it).